python/feathub/processors/spark/dataframe_builder/over_window_utils.py

```diff
@@ -0,0 +1,205 @@
+                window_descriptor, agg_descriptors, window_spec
+            )
+        )
+
+        # For rows that do not satisfy the filter predicate, set the feature col
+        # to NULL.
+        null_cols = {}
+        for descriptor in agg_descriptors:
+            null_cols[descriptor.field_name] = functions.lit(None)
+        null_feature_table = dataframe.filter(
+            ~functions.expr(window_descriptor.filter_expr)
+        ).withColumns(null_cols)
+
+        # After union, order of the row with same grouping key is not preserved. We
+        # can only preserve the order of the row with the same grouping keys and
+        # filter condition.
+        result_table = agg_table.unionAll(null_feature_table)
+    else:
+        result_table = dataframe.withColumns(
+            _get_over_window_agg_columns(
+                window_descriptor, agg_descriptors, window_spec
+            )
+        )
+    result_table.explain()
+    return result_table
+
+
+def _get_spark_window_spec(
+    over_window_descriptor: "OverWindowDescriptor",
+) -> WindowSpec:
+    if (
+        over_window_descriptor.limit is not None
+        and over_window_descriptor.window_size is not None
+    ):
+        raise FeathubException(
+            "You cannot set window_size and limit of over window at the same time."
+        )
+
+    if len(over_window_descriptor.group_by_keys) == 0:
+        window = Window.orderBy(EVENT_TIME_ATTRIBUTE_NAME)
+    else:
+        window = Window.partitionBy(*over_window_descriptor.group_by_keys).orderBy(
+            EVENT_TIME_ATTRIBUTE_NAME
+        )
+
+    if over_window_descriptor.limit is not None:
+        return window.rowsBetween(1 - over_window_descriptor.limit, Window.currentRow)
+
+    if over_window_descriptor.window_size is not None:
+        preceding = int(over_window_descriptor.window_size / timedelta(milliseconds=1))
+        return window.rangeBetween(-preceding, Window.currentRow)
+
+    return window.rowsBetween(Window.unboundedPreceding, Window.currentRow)
+
+
+def _get_over_window_agg_columns(
+    window_descriptor: "OverWindowDescriptor",
+    agg_descriptors: List["AggregationFieldDescriptor"],
+    window_spec: WindowSpec,
+) -> Dict[str, Column]:
+    return {
+        descriptor.field_name: _get_over_window_agg_column(
+            descriptor.expr,
+            window_descriptor,
+            descriptor.agg_func,
+            window_spec,
+        )
+        .cast(descriptor.field_data_type)
+        .alias(descriptor.field_name)
+        for descriptor in agg_descriptors
+    }
+
+
+def _get_over_window_agg_column(
+    expr: str,
+    over_window_descriptor: "OverWindowDescriptor",
+    agg_func: AggFunc,
+    window_spec: WindowSpec,
+) -> Column:
+
+    if (
+        over_window_descriptor.limit is not None
+        and over_window_descriptor.window_size is not None
+    ):
+        # TODO Adds aggregations on both limited and timed window
+        ...
```

Comment:
What is the meaning of this line? Should we raise an Exception here?

Reply:
I raised the Exception during constructing `WindowSpec`, which is called before this. And this line is just a marker, I will move it into the function where I raise the Exception.